(GH-2000) Deprecate bolt.yaml (#2397)
puppetlabs/bolt: documentation/upgrading_to_bolt_3.md

# Upgrading to Bolt 3.0

It's almost time for Bolt 3.0! This page contains a list of things you can do to prepare for the

> upcoming release. For an exhaustive list of the things we're changing, see [Changes coming in Bolt

Reviewer: ```suggestion
upcoming release. We'll add to this page as we near the release and implement more of the expected changes and deprecations, so if you're making the changes ahead of time, be sure to check in every couple weeks to see if we've added anything else to the instructions below. For an exhaustive list of the things we're changing, see [Changes coming in Bolt
```

Reviewer: I thought about this a bit more last night and decided we should hint at the fact that this is a living page until the release goes out. Or someone might make the changes early and miss stuff we add later.

Bolt 3.0 introduces changes to Bolt's configuration file layout and content, including
removing `bolt.yaml` and renaming several configuration options. Here's how to transition to the new
world order.

## Automated migration

For project-level configuration files you can update to use the latest Bolt best practices using our
migration command:

_\*nix shell command_

```shell
bolt project migrate
```

_PowerShell cmdlet_

```powershell
Update-BoltProject
```

## Manual migration

For user- and system-level configuration files, or if you're not comfortable having Bolt re-write
your project-level configuration files, follow these steps to migrate to the new file layout.

### User- and system-level config

1. Move any [transport configuration](bolt_transports_reference.md) to be under the
   `inventory-config` key, like so:

   ```
   inventory-config:
       ssh:
         password: hunter2!
         user: bolt
       winrm:
         password: hunter2!
         user: bolt
   ```
   Transport configuration keys are `ssh`, `winrm`, `pcp`, `local`, `docker`, `remote`, and
   `transport`.
1. Rename the following configuration options:
    - `plugin_hooks` to `plugin-hooks`

> 1. Rename `apply_settings` to `apply_settings` and move the configuration to `bolt-project.yaml`.

Reviewer 2: ```suggestion
1. Rename `apply_settings` to `apply_settings` and move the configuration to `bolt-project.yaml`
   in your project directory.
```

Just so users don't think this means to create a `bolt-project.yaml` in the user/system config directories. 😬

# Upgrading to Bolt 3.0

Welcome to Bolt 3.0! 

**Before you upgrade:**
- Take a look at the
  [Changelog](https://github.com/puppetlabs/bolt/blob/main/CHANGELOG.md).

You can view upgrade instructions for your operating system at [Installing
Bolt](bolt_installing.md).

## Migrating configuration files

Bolt 3.0 introduces changes to Bolt's configuration file layout and content, including
removing `bolt.yaml` and renaming several configuration options. Here's how to transition to the new
world order.

## Automated migration

For project-level configuration files you can update to use the latest Bolt best practices using our
migration command:

_\*nix shell command_

```shell
bolt project migrate
```

_PowerShell cmdlet_

```powershell
Update-BoltProject
```

## Manual migration

For user- and system-level configuration files, or if you're not comfortable having Bolt re-write
your project-level configuration files, follow these steps to migrate to the new file layout.

### User- and system-level config

> 

Reviewer: ```suggestion

To migrate your system- or user-level Bolt configuration, make the following changes to the relevant `bolt.yaml` file:
```
Just providing a bit of a starting point - I don't think it's clear here that we're talking about the `bolt.yaml` file.

> 1. Move any [transport configuration](bolt_transports_reference.md) to be under the

Reviewer: ```suggestion
1. Move any [transport configuration](bolt_transports_reference.md) to be under an
```

   `inventory-config` key, like so:

   ```
   inventory-config:
       ssh:
         password: hunter2!
         user: bolt
       winrm:
         password: hunter2!
         user: bolt
   ```
   Transport configuration keys are `ssh`, `winrm`, `pcp`, `local`, `docker`, `remote`, and
   `transport`.

> 1. Rename the following configuration options:
>     - `plugin_hooks` to `plugin-hooks`

Reviewer: ```suggestion
1. Rename the `plugin_hooks` configuration to `plugin-hooks`.
```

Author: We'll be adding more to this list - I know it looks a little silly now but there'll be multiple values probably by next week :grimacing:

Reviewer: Ah gotcha! No worries

> 1. Rename `apply_settings` to `apply_settings` and move the configuration to `<PROJECT
>    DIRECTORY>/bolt-project.yaml`. The default project directory is `~/.puppetlabs/bolt`.

Reviewer: ```suggestion
1. Rename `apply_settings` to `apply_settings` and move the `apply_settings` configuration to
your project directory at `<PROJECT DIRECTORY>/bolt-project.yaml`. If you'd rather use the
default Bolt project directory, place the file in `~/.puppetlabs/bolt/`.
```

# Upgrading to Bolt 3.0

Welcome to Bolt 3.0! 

**Before you upgrade:**
- Take a look at the
  [Changelog](https://github.com/puppetlabs/bolt/blob/main/CHANGELOG.md).

You can view upgrade instructions for your operating system at [Installing
Bolt](bolt_installing.md).

## Migrating configuration files

> Bolt 3.0 makes a lot of changes to Bolt's configuration file layout and content, including

Reviewer 2: ```suggestion
Bolt 3.0 makes changes to Bolt's configuration file layout and content, including
```

I think 'a lot' can be intimidating and suggests more work for the user than there might be.

Author: Good point! I actually spent a morning trying to configure vim to highlight a bunch of words and phrases like this ('very', 'really', etc.) and didn't have success - I might try again over the long weekend.

removing `bolt.yaml` and renaming several configuration options. Here's how to transition to the new
world order.

## Automated migration

For project-level configuration files you can update to use the latest Bolt best practices using our
migration command:

_\*nix shell command_

```shell
bolt project migrate
```

_PowerShell cmdlet_

```powershell
Update-BoltProject
```

## Manual migration

For user- and system-level configuration files, or if you're not comfortable having Bolt re-write
your project-level configuration files, follow these steps to migrate to the new file layout.

### User- and system-level config

1. Move any [transport configuration](bolt_transports_reference.md) to be under the
   `inventory-config` key, like so:

   ```
   inventory-config:
       ssh:
         password: hunter2!
         user: bolt
       winrm:
         password: hunter2!
         user: bolt
   ```
   Transport configuration keys are `ssh`, `winrm`, `pcp`, `local`, `docker`, `remote`, and
   `transport`.
1. Rename the following configuration options:

>     - `apply_settings` to `apply-settings`

Reviewer 2: ```suggestion
    - `apply_settings` to `apply-settings`
```

This isn't available in `bolt-defaults.yaml`. We should have a separate section that says to move this to `bolt-project.yaml`.

    - `plugin_hooks` to `plugin-hooks`

>     - `puppetfile` to `module-install`

Reviewer 2: ```suggestion
```

Not yet.

Author: Too excited :sweat_smile:

1. Rename the file to `bolt-defaults.yaml`.

### Project-level config

1. Move any [transport configuration](bolt_transports_reference.md) to the top-level `config` key of
   the inventory file. This should be at `<PROJECT DIRECTORY>/inventory.yaml`.

   ```
    targets:
      - my target

    config:
       ssh:
         password: hunter2!
         user: bolt
       winrm:
         password: hunter2!
         user: bolt
   ```
   Transport configuration keys are `ssh`, `winrm`, `pcp`, `local`, `docker`, `remote`, and
   `transport`.
1. Rename the following configuration options:
    - `apply_settings` to `apply-settings`
    - `plugin_hooks` to `plugin-hooks`

>     - `puppetfile` to `module-install`

Reviewer 2: ```suggestion
```

Not yet.
